examples/t8code_3d_dgsem/elixir_advection_nonconforming.jl

```diff
@@ -0,0 +1,113 @@
+using OrdinaryDiffEq
+using Trixi
+
+###############################################################################
+# Semidiscretization of the linear advection equation.
+
+advection_velocity = (0.2, -0.7, 0.5)
+equations = LinearScalarAdvectionEquation3D(advection_velocity)
+
+# Create DG solver with polynomial degree = 3 and (local) Lax-Friedrichs/Rusanov flux as surface flux.
+solver = DGSEM(polydeg = 3, surface_flux = flux_lax_friedrichs)
+
+coordinates_min = (-1.0, -1.0, -1.0) # minimum coordinates (min(x), min(y), min(z))
+coordinates_max = (1.0, 1.0, 1.0) # maximum coordinates (max(x), max(y), max(z))
+trees_per_dimension = (1, 1, 1)
+
+# Note that it is not necessary to use mesh polydeg lower than the solver polydeg
+# on a Cartesian mesh.
+# See https://doi.org/10.1007/s10915-018-00897-9, Section 6.
+mesh = T8codeMesh(trees_per_dimension, polydeg = 3,
+                  coordinates_min = coordinates_min, coordinates_max = coordinates_max,
+                  initial_refinement_level = 2)
+
+# Note: This is actually a `p8est_quadrant_t` which is much bigger than the
+# following struct. But we only need the first four fields for our purpose.
+struct t8_dhex_t
+    x::Int32
+    y::Int32
+    z::Int32
+    level::Int8
+    # [...] # See `p8est.h` in `p4est` for more info.
+end
+
+# Refine bottom left quadrant of each second tree to level 2
+function adapt_callback(forest,
+                        forest_from,
+                        which_tree,
+                        lelement_id,
+                        ts,
+                        is_family,
+                        num_elements,
+                        elements_ptr)::Cint
+    elements = unsafe_wrap(Array, elements_ptr, num_elements)
+    el = unsafe_load(Ptr{t8_dhex_t}(elements[1]))
+
+    if iseven(convert(Int, which_tree)) && el.x == 0 && el.y == 0 && el.z == 0 &&
+       el.level < 3
+        # return true (refine)
+        return 1
+    else
+        # return false (don't refine)
+        return 0
+    end
+end
+
+@assert(Trixi.t8_forest_is_committed(mesh.forest)!=0);
+
+# Init new forest.
+new_forest_ref = Ref{Trixi.t8_forest_t}()
+Trixi.t8_forest_init(new_forest_ref);
+new_forest = new_forest_ref[]
+
+# Check out `examples/t8_step4_partition_balance_ghost.jl` in
+# https://github.com/DLR-AMR/T8code.jl for detailed explanations.
+let set_from = C_NULL, recursive = 1, set_for_coarsening = 0, no_repartition = 0,
+    do_ghost = 1
+
+    Trixi.t8_forest_set_user_data(new_forest, C_NULL)
+    Trixi.t8_forest_set_adapt(new_forest, mesh.forest,
+                              Trixi.@t8_adapt_callback(adapt_callback), recursive)
+    Trixi.t8_forest_set_balance(new_forest, set_from, no_repartition)
+    Trixi.t8_forest_set_partition(new_forest, set_from, set_for_coarsening)
+    Trixi.t8_forest_set_ghost(new_forest, do_ghost, Trixi.T8_GHOST_FACES)
+    Trixi.t8_forest_commit(new_forest)
```

Comment:
Are these functions expected to be (really) stable API of T8code.jl?

I'm asking since we consider everything documented in our examples as stable API. Thus, if T8code.jl makes a breaking release changing this, it will also require a breaking release of Trixi.jl. We should remember this.

Reply:
I bundled these calls into an `adapt!` routine and refactored the examples accordingly. This hides the t8code API from the usual Trixi user.